users/helauth/pipeline.py

```diff
@@ -102,3 +102,61 @@ def update_user_from_profile(request: WSGIRequest, *, user: User | None = None)
         user.date_of_birth = birthday_info["birthday"]
 
     user.save()
+
+
+def migrate_user_from_tunnistamo_to_keycloak(
+    backend: TunnistamoOIDCAuth,
+    request: WSGIRequest,
+    response: OIDCResponse,
+    user: User | None = None,
+    **kwargs: Unpack[ExtraKwargs],
+) -> dict[str, Any]:
+    if user is None:
+        return {"user": user}
+
+    id_token = IDToken.from_string(response["id_token"])
+    if id_token is not None and id_token.is_ad_login and id_token.iss.endswith("helsinki-tunnistus"):
```

Comment:
Should perhaps further limit this to `@hel.fi` and `@*.hel.fi` emails (there's at least `@edu.hel.fi`), which would also prevent it being applied to the 70 or so AD users with an empty email (at least in the Tunnistamo token, not sure about Keycloak) -- in production, they seem to consist of persons with Sarastia in AD groups -- plus Eemeli -- that then need to be manually migrated for roles (and possibly reservations, if they had the Reserver role).

Reply:
Although come to think of it, in test env. we may want to test with AD guest users (with nevertheless non-empty email and if so best to enforce the email domains that can be migrated / default to none, instead of any, if not specifically configured for the environment).

Reply:
I'll add checks for invalid/empty emails.

`id_token.is_ad_login` now validates that the authentication methods reference is `helsinki_adfs`, `helsinkiad`, or `helsinkiazuread`. Isn't this enough to determine that the migration should be attempted? Or is there another `amr` for the `edu` people?

Reply:
There is currently one production AD user with an **@edu.hel.fi** email who has a staff role (unit _Reserver_).

(In theory it is possible that there are other similar users created based on AD authentication via Tunnistamo but role has never been granted, or has been removed. But I suppose that is unlikely, unless a user has used the login method by mistake. It is also possible that some of the Kasko Sarastia AD users without email in Tunnistamo token are really edu people as well, but then those users cannot be migrated anyway.)

So at least the authentication method scenario with presumably new `amr` value coming from Keycloak indeed needs to be supported (if only to identify the login method as AD for other purposes; migrating a single user's roles manually would not be a big deal in itself). Currently with Tunnistamo the amr would be still be `'helsinki_adfs'`.

As for users who have used their **@edu.hel.fi** address for `suomi.fi` authentication and profile, they can apparently keep using those profiles (and contrary to previous information, create new profiles with such emails) as long as they don't wish to authenticate with (edu) AD, in which case the profile's email would need to be changed first (but we might not know about that change if they haven't logged in as that user afterwards).

Reply:
> presumably new `amr` value coming from Keycloak indeed needs to be supported

Should be checked in staging and production, but at least on dev the `amr` is `helsinkiad` for my account, which is currently supported in `ADLoginAMR`.

Reply:
> There is currently one production AD user with an @edu.hel.fi email who has a staff role (unit Reserver).

Added `eduad` as a possible `ADLoginAMR` value to support this users migration, however if it doesn't work then they can be manually migrated since there is just one.